src/telemetry/Manager.h

```diff
@@ -50,6 +58,101 @@ class Manager
 	 */
 	virtual void InitPostScript();
 
+	/**
+	 * Instances returned by CollectMetrics().
+	 */
+	enum MetricType
+		{
+		DOUBLE_COUNTER,
+		INT_COUNTER,
+		DOUBLE_GAUGE,
+		INT_GAUGE,
+		DOUBLE_HISTOGRAM,
+		INT_HISTOGRAM,
+		};
+
+	struct CollectedMetric
+		{
+
+		CollectedMetric(Manager::MetricType metric_type,
+		                const broker::telemetry::metric_family_hdl* family,
+		                std::vector<std::string_view> label_values, std::optional<double> value,
+		                std::optional<int64_t> count_value)
+
+			: metric_type(metric_type), family(family), label_values(std::move(label_values)),
+			  value(value), count_value(count_value)
+			{
+			}
+
+		enum MetricType metric_type;
+		const broker::telemetry::metric_family_hdl* family;
+		std::vector<std::string_view> label_values;
+		std::optional<double> value;
+		std::optional<int64_t> count_value;
+
+		/**
+		 * Create a Metric record.
+		 */
+		zeek::RecordValPtr GetMetricRecord() const;
+		};
+
+	struct CollectedHistogramMetric
+		{
+		CollectedHistogramMetric(MetricType metric_type,
+		                         const broker::telemetry::metric_family_hdl* family,
+		                         std::vector<std::string_view> label_values, int64_t count,
+		                         std::vector<double> values, double sum,
+		                         std::optional<std::vector<int64_t>> count_values,
+		                         std::optional<int64_t> count_sum, std::vector<double> bounds,
+		                         std::optional<std::vector<int64_t>> count_bounds)
```

Comment:
I found it very hard to follow this code, as it's flattening a data structure and duplicates arguments for each type at the same time. This also makes it rather easy to construct a logically inconsistent object by mixing types.

Logically, a histogram is an ordered list of buckets. Why not preserve this structure?

Something along those lines:

```
template <class T>
struct Bucket {
    T upper_bound;
    int64_t count;
};

template<class T>
struct Histogram {
  T sum;
  std::vector<Bucket<T>> buckets;
};

using IntHistogram = std::vector<Histogram<int64_t>>;

using DblHistogram = std::vector<Histogram<int64_t>>;

CollectedHistogramMetric(const broker::telemetry::metric_family_hdl* family, std::vector<std::string_view> label_values, std::variant<IntHistogram, DblHistogram> data)
// ...
```

Instead of the `variant` argument, this could also have two constructors instead if you prefer that style. In any case, the `metric_type` argument should not be necessary. This is always a histogram and we should be able to figure out whether it's a `double` or `int64_t` histogram based on the arguments.

Reply:
Thanks for that!